### PR TITLE
[fsdp] feat: integrate TiledMLP for memory-efficient MLP computation

### PR DESCRIPTION
**Pull request by kevssim**
### What does this PR do?

This PR integrates TiledMLP for memory-efficient MLP computation during FSDP2 training. TiledMLP reduces peak memory usage by processing the MLP forward/backward pass in chunks (tiles), which is particularly useful for training large models with limited GPU memory.

Key features:
- Reduces peak activation memory by splitting MLP computation into configurable shards
- Compatible with FSDP2 only (not compatible with FSDP1 due to gradient computation requirements)
- Supports Llama, Qwen2, Qwen2.5, and Qwen3 model architectures
- Configurable via `tiled_mlp.enabled` and `tiled_mlp.num_shards` in model config
- Works for both actor and reference models, as well as critic models

**Note:** TiledMLP requires FSDP2 (`strategy=fsdp2`). Using it with FSDP1 will raise an error.

### Test

**Precision Test (2x H800, FSDP2, bf16, Qwen3-1.7B, num_hidden_layers=2, from_pretrained):**

| Metric | Max Diff | Mean Diff |
|--------|----------|-----------|
| Logits | 0.00e+00 | 0.00e+00 |

MLP Parameter Gradients:

| Parameter | Max Diff | Mean Diff |
|-----------|----------|-----------|
| layers.0.mlp.down_proj.weight | 9.77e-04 | 1.42e-06 |
| layers.0.mlp.gate_proj.weight | 9.77e-04 | 7.41e-07 |
| layers.0.mlp.up_proj.weight | 9.77e-04 | 1.04e-06 |
| layers.1.mlp.down_proj.weight | 1.95e-03 | 1.20e-06 |
| layers.1.mlp.gate_proj.weight | 3.91e-03 | 4.77e-07 |
| layers.1.mlp.up_proj.weight | 3.91e-03 | 1.12e-06 |

The forward pass (logits) is identical. Gradient differences are within expected numerical precision for bf16 tiled computation.

**Performance Test:**

#### 1. FSDP2 + GPU (2x NVIDIA H800)

Env: 2x NVIDIA H800, FSDP2, bf16, batch_size=1, num_shards=4, hidden_size=4096, intermediate_size=12288

| Seq Length | Regular (ms) | Tiled (ms) | Speed | Regular (MB) | Tiled (MB) | Memory |
|------------|--------------|------------|-------|--------------|------------|--------|
| 2048       | 18.59        | 23.51      | +26%  | 1216.00      | 1444.00    | +19%   |
| 4096       | 22.43        | 31.57      | +41%  | 1504.00      | 1576.00    | +5%    |
| 8192       | 33.11        | 46.01      | +39%  | 2080.00      | 1840.00    | **-12%**   |
| 16384      | 54.06        | 76.76      | +42%  | 3424.00      | 2368.00    | **-31%**   |
| 32768      | 103.93       | 136.67     | +31%  | 6112.00      | 3456.00    | **-43%**   |
| 65536      | 190.89       | 258.67     | +36%  | 11488.00     | 5696.00    | **-50%**   |
| 131072     | 391.50       | 501.15     | +28%  | 22240.00     | 10176.00   | **-54%**   |

#### 2. FSDP2 + NPU (910B)

Env: FSDP2, bf16, NPU

| Seq Length | Regular (ms) | Tiled (ms) | Speed | Regular (MB) | Tiled (MB) | Memory |
|------------|--------------|------------|-------|--------------|------------|--------|
| 2048       | 54.41        | 46.26      | -15%  | 1152.01      | 1460.02    | +27%   |
| 4096       | 65.69        | 81.05      | +23%  | 1460.01      | 1576.01    | +8%    |
| 8192       | 80.46        | 98.85      | +23%  | 2036.01      | 1808.01    | **-11%**   |
| 16384      | 108.22       | 141.93     | +31%  | 3360.01      | 2324.01    | **-31%**   |
| 32768      | 177.25       | 228.90     | +29%  | 6048.01      | 3392.01    | **-44%**   |
| 65536      | 313.80       | 419.55     | +34%  | 11424.01     | 5632.01    | **-51%**   |
| 131072     | 601.27       | 791.40     | +32%  | 22176.01     | 10112.01   | **-54%**   |

**Note**: Speed + means tiled is slower, Memory - means tiled uses less memory

**Summary:** TiledMLP trades ~30-40% speed for significant memory savings. At longer sequences (≥8192), memory reduction reaches 50%+, making it ideal for long-context training scenarios where memory is the bottleneck.

### API and Usage Example

Enable TiledMLP in your training config:

```yaml
actor_rollout_ref:
  model:
    tiled_mlp:
      enabled: true
      num_shards: 4  # Higher values = lower memory, slightly slower
```

Or via command line:

```bash
python3 -m verl.trainer.main_ppo \
    actor_rollout_ref.model.tiled_mlp.enabled=True \
    actor_rollout_ref.model.tiled_mlp.num_shards=4 \
    actor_rollout_ref.actor.strategy=fsdp2 \
    ...
```

For critic model:

```yaml
critic:
  model:
    tiled_mlp:
      enabled: true
      num_shards: 4
```

### Design & Code Changes

**Core Implementation:**
- `verl/models/transformers/tiled_mlp.py`: New file containing:
  - `GradientAccumulator`: Manages gradient accumulation across shards during backward pass
  - `TiledMLP`: Custom autograd function that processes MLP forward/backward in tiles
  - `apply_tiled_mlp_monkey_patch()`: Patches transformers MLP classes

**Integration:**
- `verl/models/transformers/monkey_patch.py`:
  - Added `use_tiled_mlp` and `tiled_mlp_shards` parameters to `apply_monkey_patch()`
- `verl/workers/fsdp_workers.py`:
  - Pass TiledMLP config to `apply_monkey_patch()`
  - Added validation to ensure FSDP2 is used (TiledMLP requires FSDP2)

**Configuration:**
- `verl/workers/config/model.py`: Added `tiled_mlp` config field
- `verl/workers/config/critic.py`: Added `tiled_mlp` config field for critic
- `verl/trainer/config/model/hf_model.yaml`: Added TiledMLP config with documentation
- `verl/trainer/config/critic/dp_critic.yaml`: Added TiledMLP config for critic

**Supported Models:**
- LlamaMLP (llama)
- Qwen2MLP (qwen2, qwen2_5)
- Qwen3MLP (qwen3)

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why.
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ).

### CODE REVIEW

File: verl/workers/fsdp_workers.py [hunk 1]
```diff
@@ -1276,6 +1308,23 @@ def _build_critic_model_optimizer(self, config):
             use_meta_tensor=not critic_model_config.tie_word_embeddings, mesh=self.device_mesh
         )
 
+        # Apply TiledMLP monkey patch BEFORE model instantiation for memory-efficient MLP
```
Comment (wuxibin89) [inline verl/workers/fsdp_workers.py]: We better hide all monkey patch under a single endpoint: `apply_monkey_patch`
- ulyssess sp
- use_fused_kernels
- tiled mlp
- ...
